Feat/events

Finnick223 opened the pull request.

Finnick223 committed:
feat: event CRUD

- Add model, migrations, controller, seeder, routes and policy for calendar events feature

Finnick223 committed:
PHP Linting (Pint)

Finnick223 committed:
feat: create Resource for EventController index

Finnick223 committed:
Merge remote-tracking branch 'origin/feat/events' into feat/events

Finnick223 committed:
PHP Linting (Pint)

Finnick223 committed:
feat: wrap store and update event with  EventResource

Finnick223 committed:
Merge remote-tracking branch 'origin/feat/events' into feat/events

Finnick223 committed:
Merge remote-tracking branch 'origin/feat/handle-permission-frontend' into feat/events

Finnick223 committed:
feat: split event logic between Instructor and Driver

- add new relationships eg. instructor has many drivers
- create separate controller for management events by instructor and get events by driver
- add route for listing drivers of instructor
- prefix event routes with roles
- make migration foreignId to users table
- embed driver details to api response for instructor
- validate if driver belongs to instructor when storing, updating or deleting data

Finnick223 committed:
PHP Linting (Pint)

marioooo0o commented on app/Http/Controllers/InstructorEventController.php:
```diff
@@ -0,0 +1,97 @@
+
+        $driverIds = $instructor->drivers->pluck('id');
+
+        $events = Event::whereIn('user_id', $driverIds)->with('driver')->get();
```
używajmy whereIntegerInRaw to wpływa na wydajność zapytania

marioooo0o commented on app/Http/Controllers/InstructorEventController.php:
```diff
@@ -0,0 +1,97 @@
+     */
+    public function store(Request $request)
+    {
+        $validated = $request->validate([
+            'driver_id' => 'required|exists:users,id',
+            'title' => 'required|string|max:255',
+            'start' => 'required|date',
+            'end' => 'required|date',
+        ]);
```
zrób custom Request Class i tam umieść walidacje

marioooo0o replied:
oraz Value Object

marioooo0o commented on app/Http/Controllers/InstructorEventController.php:
```diff
@@ -0,0 +1,97 @@
+    public function store(Request $request)
+    {
+        $validated = $request->validate([
+            'driver_id' => 'required|exists:users,id',
```
dodaj integer

marioooo0o commented on app/Http/Controllers/InstructorEventController.php:
```diff
@@ -0,0 +1,97 @@
+            'driver_id' => 'required|exists:users,id',
+            'title' => 'required|string|max:255',
+            'start' => 'required|date',
+            'end' => 'required|date',
```
warto dodać walidację która nie pozwala na ustawienie daty końcowej która jest przed datą startu

marioooo0o commented on app/Http/Controllers/InstructorEventController.php:
```diff
@@ -0,0 +1,97 @@
+            return response()->json(['message' => 'Unauthorized event'], 403);
+        }
+
+        $validated = $request->validate([
+            'title' => 'required|string|max:255',
+            'start' => 'required|date',
+            'end' => 'required|date',
+        ]);
```
tak samo umieść w Request class

marioooo0o commented on app/Http/Resources/DriverEventResource.php:
```diff
@@ -0,0 +1,24 @@
+        return [
+            'id' => $this->id,
+            'title' => $this->title,
+            'start' => $this->start,
+            'end' => $this->end,
```
Nie ma potrzeby tutaj zwrócić danych scastowanych?

Finnick223 committed:
feat: create custom request class to extract validation and authorization logic from controller

Finnick223 committed:
Merge remote-tracking branch 'origin/feat/events' into feat/events

# Conflicts:
#	app/Http/Controllers/InstructorEventController.php

Finnick223 committed:
PHP Linting (Pint)

Finnick223 committed:
feat: implement Value Object when storing events and improve event edit and delete policy

Finnick223 committed:
Merge remote-tracking branch 'origin/feat/events' into feat/events

# Conflicts:
#	app/Http/Controllers/InstructorEventController.php
#	app/Http/Requests/StoreInstructorEventRequest.php

Finnick223 committed:
PHP Linting (Pint)

marioooo0o commented on app/Http/Controllers/InstructorEventController.php:
```diff
@@ -0,0 +1,83 @@
+     */
+    public function store(StoreInstructorEventRequest $request)
+    {
+        $validated = $request->validated();
+        $dateRange = $request->getEventDateRange();
+        $instructor = $request->user();
+        $driver = $instructor->drivers()->findOrFail($validated['driver_id']);
+
+        $event = $driver->events()->create([
+            'title' => $validated['title'],
+            'start' => $dateRange->start,
+            'end' => $dateRange->end,
+        ]);
+        $event->load('driver');
+
+        return response()->json(new InstructorEventResource($event), 201);
```
logika biznesowa w Service i korzystaj z value objects

marioooo0o replied:
Kody HTTP ze stałej Response Http Foundation

marioooo0o commented on app/Http/Controllers/InstructorEventController.php:
```diff
@@ -0,0 +1,83 @@
+     */
+    public function update(UpdateInstructorEventRequest $request, Event $event)
+    {
+        $validated = $request->validated();
+
+        $event->update([
+            'title' => $validated['title'],
+            'start' => $validated['start'],
+            'end' => $validated['end'],
+        ]);
+        $event->load('driver');
+
+        return response()->json(new InstructorEventResource($event));
```
podobnie tutaj

marioooo0o commented on app/Http/Requests/UpdateInstructorEventRequest.php:
dodaj value object

marioooo0o commented on app/ValueObjects/EventDateRange.php:
```diff
@@ -0,0 +1,31 @@
+        $startDate = new DateTimeImmutable($start);
+        $endDate = new DateTimeImmutable($end);
+
+        if ($endDate <= $startDate) {
+            throw new InvalidArgumentException('The end date must be after the start date.');
+        }
```
tutaj masz +

Finnick223 committed:
refactor: improve event management architecture and code structure

- Separate business logic from controllers into EventService
- Change value object to EventDetails for better data encapsulation
- Update InstructorEventController to use EventService
- Refactor request classes to include getEventDetails method
- Improve code organization, readability and reduce controller complexity

Finnick223 committed:
PHP Linting (Pint)

Finnick223 committed:
feat: add status of event and create accept/reject functionality for admin

- add status to events
- make enum for status values
- create AdminEventController.php to return pending events, accept or reject for admin only
- extract logic to AdminEventService.php
- create SSE endpoint for realtime frontend pending events panel

Finnick223 committed:
PHP Linting (Pint)

Finnick223 committed:
fix: cors issue on StreamEventService.php

Finnick223 committed:
Merge remote-tracking branch 'origin/feat/events' into feat/events

# Conflicts:
#	app/Services/StreamEventService.php

Finnick223 committed:
PHP Linting (Pint)

Finnick223 committed:
refactor: admin routes in separate file

Finnick223 committed:
Merge remote-tracking branch 'origin/feat/events' into feat/events

Finnick223 committed:
PHP Linting (Pint)

marioooo0o commented on app/Http/Requests/StoreInstructorEventRequest.php:
```diff
@@ -0,0 +1,38 @@
+    public function rules(): array
+    {
+        return [
+            'driver_id' => 'required|integer|exists:users,id',
```
sprawdzić czy user ma rolę driver

marioooo0o commented on app/Http/Resources/InstructorEventResource.php:
```diff
@@ -0,0 +1,26 @@
+            'id' => $this->id,
+            'driver' => new DriverResource($this->whenLoaded('driver')),
+            'title' => $this->title,
+            'start' => $this->start,
+            'end' => $this->end,
```
prawdopodobnie będziesz potrzebował dodać na końcu ->toDateTimeString()

marioooo0o commented on app/Http/Controllers/AdminEventController.php:
```diff
@@ -0,0 +1,36 @@
+        return response()->json($this->eventService->getPendingEvents(), Response::HTTP_OK);
+    }
+
+    public function acceptEvent($id)
```
Zastosujmodel binding https://laravel.com/docs/11.x/routing#route-model-binding

marioooo0o commented on app/Http/Controllers/AdminEventController.php:
```diff
@@ -0,0 +1,36 @@
+        return response()->json(['message' => 'Event accepted'], Response::HTTP_OK);
+    }
+
+    public function rejectEvent($id)
```
tak samo tutaj model binding

marioooo0o commented on app/Http/Controllers/AdminEventController.php:
```diff
@@ -0,0 +1,36 @@
+
+    public function getPendingEvents()
+    {
+        return response()->json($this->eventService->getPendingEvents(), Response::HTTP_OK);
```
resource

marioooo0o commented on app/Http/Controllers/AdminEventController.php:
```diff
@@ -0,0 +1,36 @@
+    {
+        $event = $this->eventService->acceptEvent($id);
+
+        return response()->json(['message' => 'Event accepted'], Response::HTTP_OK);
```
.

marioooo0o commented on app/Http/Controllers/AdminEventController.php:
```diff
@@ -0,0 +1,36 @@
+    {
+        $this->eventService->rejectEvent($id);
+
+        return response()->json(['message' => 'Event rejected'], Response::HTTP_OK);
```
.

marioooo0o commented on routes/api/owner.php:
```diff
@@ -0,0 +1,16 @@
+use App\Http\Controllers\AdminEventController;
+use Illuminate\Support\Facades\Route;
+
+$admin = RolesEnum::OWNER->value;
```
używasz gdzieś tej zmiennej?

Finnick223 replied:
Dodałem dla spójności, bo w driver  i instructor również takie są

marioooo0o reviewed (changes requested):
Dodałem Ci drobner uwagi, poza tym co mówiliśmy na spotkaniu

Finnick223 committed:
feat: add model binding, pagination, AdminEventResource.php instead of response, remove stream related code and small fixes

Finnick223 committed:
PHP Linting (Pint)

Finnick223 committed:
feat: include instructor_id to event table and some improvements

- create necessary migration
- rename user_id to driver_id to make it easier to read
- create InstructorResource
- improve naming

Finnick223 committed:
PHP Linting (Pint)

Finnick223 committed:
fix: frontend pagination component need paginate instead of simplePaginate

Finnick223 committed:
Merge remote-tracking branch 'origin/feat/events' into feat/events

Finnick223 committed:
feat: create instructor availability management

Implement backend logic for getting and storing instructor availability. Each new submission replaces old availability to align with frontend design, where users set availability for the current and next week in bulk. Separate POST and UPDATE endpoints seems unnecessary.

Finnick223 committed:
PHP Linting (Pint)

Finnick223 committed:
feat: create driver availability management

- Implement backend logic for getting instructor availability as driver and create pending events based on this availability.
- Add status enum
- migration with status column

Finnick223 committed:
PHP Linting (Pint)

Finnick223 committed:
fix: missing status causing error

Finnick223 committed:
feat: add status into InstructorAvailabilityResource.php

status is needed to handle booked events on frontend availability panel for instructor

Finnick223 committed:
fix: rejected event timeslot is available again

Finnick223 committed:
merge dev into branch and resolve conflicts

Finnick223 committed:
PHP Linting (Pint)
